noise_corrector.py

```diff
@@ -1,66 +1,204 @@
+T = TypeVar('T')
+
+
+class ParamConfig:
+  def __init__(self,
+               randomized_reponse_rate,
+               trigger_data_cardinality,
+               max_reports,
+               max_windows,
+               name):
+    self.rr_rate = randomized_reponse_rate
+    self.data_cardinality = trigger_data_cardinality
+    self.max_reports = max_reports
+    self.max_windows = max_windows
+    self.name = name
+
+
+NAV_PARAMS = ParamConfig(.5, 8, 3, 3, 'navigation')
+EVENT_PARAMS = ParamConfig(.0000025, 2, 1, 1, 'event')
+
+DEFAULT_EXPIRY = 60 * 60 * 24 * 30
+
+class OutputEnumeration:
+  # Class encapsulating logic which represents the logical output of the API
+  # per source event, for the purposes of debiasing the randomized response
+  # privacy mechanism.
+  def __init__(self, report_tuple: OutputConfig, params: ParamConfig):
+    '''report_tuple must be in the form ((window_index, trigger_data),...)'''
+    self.output: OutputConfig = tuple(sorted(report_tuple))
+    self.params = params    
+
+  def __str__(self):
+    def report_to_string(r):
+      if r is None:
+        return 'No report'
+      return {'window': r[0], 'data': r[1]}
+    return json.dumps([report_to_string(r) for r in self.output])
+
+  def __repr__(self): return self.output.__repr__()
+  def __hash__(self): return self.output.__hash__()
+
+  # Needed to support instances in dictionaries
+  def __eq__(self, other): return self.output.__eq__(other.output)
+  def __lt__(self, other): return self.output.__lt__(other.output)
+
+  @classmethod
+  def create_null(cls, params: ParamConfig):
+    return cls((), params)
+
+  @classmethod
+  def create_from_data(cls, source_with_reports):
+    '''Create an output directly from structured data
+
+    source_with_reports: list of (source, [report, ...]) tuples.
+    See get_conversion_counts for expected structure for sources and reports.
+    '''
+    source, reports = source_with_reports
+    params = EVENT_PARAMS if source['source_type'] == 'event' else NAV_PARAMS
+    max_reports = params.max_reports
+    assert len(reports) <= max_reports
+
+    def parse_report(source, report: dict) -> Tuple[int, int]:
+      window_index = OutputEnumeration._get_window_index_for_report(
+          source, report)
+      trigger_data = int(report['report']['trigger_data'])
+      return (window_index, trigger_data)
+
+    return cls(tuple(parse_report(source, r) for r in reports), params)
+
+  @classmethod
+  def generate_all(cls, params: ParamConfig):
+    all_single_reports = list(itertools.product(range(params.max_windows),
+                                                range(params.data_cardinality)))
+    all_reports_ordered = []
+    for r in range(params.max_reports + 1):
+      reports = itertools.product(*(r * [all_single_reports]))
+      all_reports_ordered.extend(list(reports))
+    # Simulate shuffling within a reporting window by sorting and deduping
+    deduped = set([tuple(sorted(x)) for x in all_reports_ordered])
```

Comment:
This seems a bit inefficient given that we may generate some state multiple times (and then dedupe). Wouldn't it be possible to just directly generate the reports that we want without duplication? For example, I had written some code below a while back (the parameter names are different but the logic should be similar). Would this work?

```
def all_possible_reports(
    data_cardinality, max_windows, max_reports, index=0, report_so_far=[]):
  yield report_so_far
  if max_reports == 0:
    return
  for next_index in range(index, data_cardinality * max_windows):
    window, data = divmod(next_index, data_cardinality)
    report_so_far.append({"window": window, "data": data})
    yield from all_possible_reports(
        data_cardinality,
        max_windows,
        max_reports - 1,
        index=next_index,
        report_so_far=report_so_far)
    report_so_far.pop()
```